File: packages/typescript/server/src/permissions.ts
/**
 * Permission discovery — query what you can do before trying it.
 */

import type {
  DelegationToken as DelegationTokenType,
  CapabilityDeclaration as CapabilityDeclarationType,
} from "@anip-dev/core";

export interface PermissionResult {
  available: Array<{
    capability: string;
    scope_match: string;
    constraints: Record<string, unknown>;
  }>;
  restricted: Array<{
    capability: string;
    reason: string;
    reason_type: string;
    grantable_by: string;
    unmet_token_requirements?: string[];
    resolution_hint?: string | null;
  }>;
  denied: Array<{
    capability: string;
    reason: string;
    reason_type: string;
  }>;
}

/**
 * Given a delegation token and a map of capabilities, return what the agent
 * can and cannot do.
 */
export function discoverPermissions(
  token: DelegationTokenType,
  capabilities: Record<string, CapabilityDeclarationType>,
): PermissionResult {
  const available: PermissionResult["available"] = [];
  const restricted: PermissionResult["restricted"] = [];
  const denied: PermissionResult["denied"] = [];

  const tokenScopeBases: Array<[string, string]> = token.scope.map((s) => [
    s.split(":")[0],
    s,
  ]);
  const rootPrincipal = token.root_principal ?? token.issuer;

  for (const [name, cap] of Object.entries(capabilities)) {
    const requiredScopes = cap.minimum_scope;
    const matchedScopeStrs: string[] = [];
    const missing: string[] = [];

    for (const required of requiredScopes) {
      let matchedFull: string | null = null;
      for (const [scopeBase, fullScope] of tokenScopeBases) {
        if (
          scopeBase === required ||
          required.startsWith(scopeBase + ".")
        ) {
          matchedFull = fullScope;
          break;
        }
      }
      if (matchedFull !== null) {
        matchedScopeStrs.push(matchedFull);
      } else {
        missing.push(required);
      }
    }

    if (missing.length === 0) {
      // Check token-evaluable control requirements
      const controlRequirements = (cap as any).control_requirements ?? [];
      const unmet: string[] = [];
      const constraintsObj = token.constraints as Record<string, unknown> | undefined;
      for (const req of controlRequirements as Array<Record<string, unknown>>) {
        const reqType = req.type as string;
        if (reqType === "cost_ceiling" && (!constraintsObj || !constraintsObj.budget)) {
          unmet.push("cost_ceiling");
        } else if (reqType === "stronger_delegation_required") {
          const tokenHasExplicitBinding = (
            token.purpose !== null &&
            token.purpose !== undefined &&
            token.purpose.capability === name
          );
          if (!tokenHasExplicitBinding) {
            unmet.push("stronger_delegation_required");
          }
        }
      }

      if (
        unmet.length > 0 &&
        controlRequirements.some(
          (r: Record<string, unknown>) =>
            r.enforcement === "reject" && unmet.includes(r.type as string),
        )
      ) {
        const ctrlResolutionHint = unmet.includes("cost_ceiling")
          ? "request_budget_bound_delegation"
          : "request_capability_binding";
        restricted.push({
          capability: name,
          reason: `missing control requirements: ${unmet.join(", ")}`,
          reason_type: "unmet_control_requirement",
          grantable_by: rootPrincipal,
          unmet_token_requirements: unmet,
          resolution_hint: ctrlResolutionHint,
        });
        continue;
      }

      const constraints: Record<string, unknown> = {};
      for (const scopeStr of matchedScopeStrs) {
        if (scopeStr.includes(":max_$")) {
          const maxBudget = parseFloat(scopeStr.split(":max_$")[1]);
          constraints.budget_remaining = maxBudget;
          constraints.currency = "USD";
        }
      }
      // Include constraints-level budget info if present
      if (constraintsObj?.budget) {
        const budget = constraintsObj.budget as Record<string, unknown>;
        constraints.budget_remaining = budget.max_amount;
        constraints.currency = budget.currency;
      }
      available.push({
        capability: name,
        scope_match: matchedScopeStrs.join(", "),
        constraints,
      });
    } else if (missing.some((s) => s.startsWith("admin."))) {
      denied.push({
        capability: name,
        reason: "requires admin principal",
        reason_type: "non_delegable",
      });
    } else {
      restricted.push({
        capability: name,
        reason: `delegation chain lacks scope(s): ${missing.join(", ")}`,
        reason_type: "insufficient_scope",
        grantable_by: rootPrincipal,
        resolution_hint: "request_broader_scope",
      });
    }
  }

  return { available, restricted, denied };
}

File: packages/python/anip-server/src/anip_server/permissions.py
"""Permission discovery — query what you can do before trying it."""
from __future__ import annotations

from anip_core import (
    AvailableCapability,
    CapabilityDeclaration,
    DelegationToken,
    DeniedCapability,
    PermissionResponse,
    RestrictedCapability,
)


def discover_permissions(
    token: DelegationToken,
    capabilities: dict[str, CapabilityDeclaration],
) -> PermissionResponse:
    """Given a delegation token, return what the agent can and can't do."""
    available: list[AvailableCapability] = []
    restricted: list[RestrictedCapability] = []
    denied: list[DeniedCapability] = []

    token_scope_bases = [(s.split(":")[0], s) for s in token.scope]
    root_principal = token.root_principal or token.issuer
    constraints_obj = token.constraints

    for name, cap in capabilities.items():
        required_scopes = cap.minimum_scope
        matched_scope_strs: list[str] = []
        missing: list[str] = []
        for required in required_scopes:
            matched_full = None
            for scope_base, full_scope in token_scope_bases:
                if scope_base == required or required.startswith(scope_base + "."):
                    matched_full = full_scope
                    break
            if matched_full is not None:
                matched_scope_strs.append(matched_full)
            else:
                missing.append(required)

        if missing:
            if len(missing) == len(required_scopes):
                # No scope overlap at all — completely inaccessible
                denied.append(
                    DeniedCapability(
                        capability=name,
                        reason=f"delegation chain lacks all required scope(s): {', '.join(missing)}",
                        reason_type="insufficient_scope",
                    )
                )
            else:
                restricted.append(
                    RestrictedCapability(
                        capability=name,
                        reason=f"delegation chain lacks scope(s): {', '.join(missing)}",
                        reason_type="insufficient_scope",
                        grantable_by=root_principal,
                        resolution_hint="request_broader_scope",
                    )
                )
            continue

        # Scope matched — check token-evaluable control requirements
        unmet: list[str] = []
        for req in cap.control_requirements:
            if req.type == "cost_ceiling" and (not constraints_obj or not constraints_obj.budget):
                unmet.append("cost_ceiling")
            elif req.type == "stronger_delegation_required":
                # Check if token has explicit capability binding via purpose
                token_has_explicit_binding = (
                    token.purpose is not None
                    and token.purpose.capability == name
                )
                if not token_has_explicit_binding:
                    unmet.append("stronger_delegation_required")

        if unmet and any(
            r.enforcement == "reject"
            for r in cap.control_requirements
            if r.type in unmet
        ):
            # Determine the most specific resolution hint based on unmet requirements
            if "cost_ceiling" in unmet:
                ctrl_resolution_hint = "request_budget_bound_delegation"
            else:
                ctrl_resolution_hint = "request_capability_binding"
            restricted.append(
                RestrictedCapability(
                    capability=name,
                    reason=f"missing control requirements: {', '.join(unmet)}",
                    reason_type="unmet_control_requirement",
                    grantable_by=root_principal,
                    unmet_token_requirements=unmet,
                    resolution_hint=ctrl_resolution_hint,
                )
            )
            continue

        # All checks passed — capability is available
        constraints: dict = {}
        for scope_str in matched_scope_strs:
            if ":max_$" in scope_str:
                max_budget = float(scope_str.split(":max_$")[1])
                constraints["budget_remaining"] = max_budget
                constraints["currency"] = "USD"
        # Include constraints-level budget info if present
        if constraints_obj and constraints_obj.budget:
            constraints["budget_remaining"] = constraints_obj.budget.max_amount
            constraints["currency"] = constraints_obj.budget.currency
        available.append(
            AvailableCapability(
                capability=name,
                scope_match=", ".join(matched_scope_strs),
                constraints=constraints,
            )
        )

    return PermissionResponse(
        available=available, restricted=restricted, denied=denied
    )
